core/src/main/java/hudson/PluginWrapper.java

```diff
@@ -524,20 +559,71 @@ public boolean hasLicensesXml() {
      *             thrown if one or several mandatory dependencies doesn't exists.
      */
     /*package*/ void resolvePluginDependencies() throws IOException {
-        List<Dependency> missingDependencies = new ArrayList<>();
+        if (ENABLE_PLUGIN_DEPENDENCIES_VERSION_CHECK) {
+            String requiredCoreVersion = getRequiredCoreVersion();
+            if (requiredCoreVersion == null) {
+                LOGGER.warning(shortName + " doesn't declare required core version.");
+            } else {
+                VersionNumber actualVersion = Jenkins.getVersion();
+                if (actualVersion.isOlderThan(new VersionNumber(requiredCoreVersion))) {
+                    dependencyErrors.add(Messages.PluginWrapper_obsoleteCore(Jenkins.getVersion().toString(), requiredCoreVersion));
+                }
+            }
+        }
         // make sure dependencies exist
         for (Dependency d : dependencies) {
-            if (parent.getPlugin(d.shortName) == null)
-                missingDependencies.add(d);
-        }
-        if (!missingDependencies.isEmpty())
-            throw new MissingDependencyException(this.shortName, missingDependencies);
+            PluginWrapper dependency = parent.getPlugin(d.shortName);
+            if (dependency == null) {
+                PluginWrapper failedDependency = NOTICE.getPlugin(d.shortName);
+                if (failedDependency != null) {
+                    dependencyErrors.add(Messages.PluginWrapper_failed_to_load_dependency(failedDependency.getLongName(), d.version));
+                    break;
+                } else {
+                    dependencyErrors.add(Messages.PluginWrapper_missing(d.shortName, d.version));
+                }
+            } else {
+                if (dependency.isActive()) {
+                    if (isDependencyObsolete(d, dependency)) {
+                        dependencyErrors.add(Messages.PluginWrapper_obsolete(dependency.getLongName(), dependency.getVersion(), d.version));
+                    }
+                } else {
+                    if (isDependencyObsolete(d, dependency)) {
+                        dependencyErrors.add(Messages.PluginWrapper_disabledAndObsolete(dependency.getLongName(), dependency.getVersion(), d.version));
+                    } else {
+                        dependencyErrors.add(Messages.PluginWrapper_disabled(dependency.getLongName()));
+                    }
+                }
 
+            }
+        }
         // add the optional dependencies that exists
         for (Dependency d : optionalDependencies) {
-            if (parent.getPlugin(d.shortName) != null)
-                dependencies.add(d);
+            PluginWrapper dependency = parent.getPlugin(d.shortName);
```

Comment:
Is this going to cause an infinite loop if the actual installed version of `d` has dependencies including this plugin.

It may be ok, but this is my fear